### ISSUE TITLE
Extracting data from tables

### ISSUE THREAD

**Issue by BeebBenjamin** [python-openxml/python-docx#13]
Hi,

Is it possible to extract data from tables using the docx module?

It would be useful to have more examples with regards to learning to use this library.

Regards,

Ben

**Comment by divyaiyer** [comment]
Is there any way I can extract tables with a custom name I had given in my template? 
Use case : Parsing a docx file and storing values in to database, docx contains lot of tables, but I want to extract only the tables with the style name 'docReader_Tables'. Is this possible? Thanks in advance.

**Comment by scanny** [comment]
Hi @divyaiyer, you might want to post this one on StackOverflow if you have a login there. It's more of a support question than a code issue.

I do think you'll need to provide a bit more detail.

Off the top of my head, tables have a different type of style than say paragraphs, if I remember correctly, and are specified by a GUID rather than a name. So that part doesn't make sense to me. The tables are available in Document.tables, so you can iterate through them that way once you have the test worked out to identify the one you want.

**Comment by divyaiyer** [comment]
Thnx for your reply @scanny . I shall post on stackOverflow, I do have an account. Ya my case is bit tricky, as I have a particular paragraph heading and I want to fetch all tables comes under that heading and store in database. Thnx

**Comment by scanny** [comment]
If your test for inclusion is following a heading paragraph, you'll need to iterate over both paragraphs and tables in document order. The document.iter_block_level_items() method is designed for that (although it hasn't been officially implemented yet.

You may want to check out this issue that has details on that.
https://github.com/python-openxml/python-docx/issues/40

**Comment by divyaiyer** [comment]
Thanks @scanny , . I tried that program, I get the blocks as Paragraph and tables, what I need exactly is I have a paragraph heading with a custom style, if I pass that to a method, that should return me all headings, sub heading and table under that. 

Ex: 

1.1. 	Heading 1 (defined cutsom style)
1.1.1. sub Heading 1(defined cutsom style)
1.1.1.1. Child of sub heading (defined cutsom style)
     ----
<Table>
1.2. 	Heading 2 (defined cutsom style)
1.2.1. sub Heading 1 (defined cutsom style)
1.2.1.1. Child of sub heading (defined cutsom style)
     ----
<Table>

**Comment by scanny** [comment]
How were you thinking to approach the problem? It sounds like you've successfully identified your "start" marker. Have you thought about how to handle the end marker? Like when to stop and return what you've found?

**Comment by divyaiyer** [comment]
@scanny , ya I know end marker is not easy. Currently I read tables and looks for row -> columns and I am validating table using first column name what I have defined. Will be parsing only such tables and once it reaches end of the document, it gets returns. While parsing I am pushing them to DB. I know this is not very good way of dealing my requirement, searching more how can I reach best solution. Thnx

**Comment by nawuy** [comment]
@scanny I tried your code:
for table in tables:
    assignments.write(table)
    for row in table.rows:
        for cell in row.cells:
            for paragraph in cell.paragraphs:
                 file.write(paragraph.text)  # I am storing the data in a txt file

My output is repeated an X amount of times, in other words when I open my txt file and I look for a unique "data" I notice that this "data" is repeated different times in the txt file. ( 6 to 8 times, depending on the cell)

**Comment by scanny** [comment]
@nawuy I believe that row.cells will return a "merged" cell multiple times, once for each cell that is merged into it. Could that be your problem?

**Comment by lxj0276** [comment]
when I parsed my docx file,I found it lost data  when I parse tables.
my method it's like:
`document = Document(path_to_your_docx)
tables = document.tables
for table in tables:
    for row in table.rows:
        for cell in row.cells:
            for paragraph in cell.paragraphs:
                print(paragraph.text)`

I found some data become empty string.

**Comment by movinghands** [comment]
@scanny This is the problem I'm facing when I'm using the library. How should I go about fixing it?

**Comment by scanny** [comment]
@movinghands Sounds like a question for StackOverflow.

**Comment by Aviru** [comment]
Hi,
I am using python-docx to read tables from docx file. But different docx files contain different table structure i.e. in some docx files the table heads are top aligned and in some files table heads are left aligned. I am unable to get the alignment/position of table heads, so when I iterate through table and  try to fetch the table data in key value pair, the data is inappropriate as the keys and values are inappropriate.
I have attached the images for table head alignment.
![left_aligned_table_head](https://user-images.githubusercontent.com/9887906/40905645-382f8e3e-67fc-11e8-9040-f93284896618.png)
![top_aligned_table_head](https://user-images.githubusercontent.com/9887906/40905646-38626c5a-67fc-11e8-810f-96b2c3ef9e55.png)



Please help me.

Thanks & Regards,
Aviru Bhattacharjee

**Comment by nawuy** [comment]
Can we see your code? Also, if I were you I would ask this question on Stackoverflow as well :)

**Comment by Aviru** [comment]
def convertDocxToText(path):
    doc = Document(path)
    fullText = []

    for block in iter_block_items(doc):
        if isinstance(block, Paragraph):
            #.... This part is for paragraph.
       
       elif isinstance(block, Table):
            for i, row in enumerate(block.rows):
                text = (cell.text for cell in row.cells)
              
                  #.... This part is for Table.
                 
                # for col in block.columns:
                #     for colCel in col.cells:
                #         print(colCel.text)
                #
                # for cell in row.cells:
                #     for paragraph in cell.paragraphs:
                #         print(paragraph.text)

                # Establish the mapping based on the first row
                # headers; these will become the keys of our dictionary
                # if i == 0:
                # tpl = tuple(text)
                # keys = tpl[0]  # tuple(text)
                # continue
                # Construct a dictionary for this row, mapping
                # keys to values for this row

                if i == 0:
                    keys = tuple(text)
                    continue

                row_data = dict(zip(keys, text))
                print(str(row_data))
                fullText.append(str(row_data))
                # for key, val in row_data.items():
                #     print(key + ": " + val)
                #     strKeyVal = key + ": " + val
                #     fullText.append(strKeyVal)

                #     for paragraph in cell.paragraphs:
                #         fullText.append(paragraph.text)
                # return ("\t".join(fullText))
    return '\n'.join(fullText)


def iter_headings(paragraphs):
    for paragraph in paragraphs:
        if paragraph.style.name.startswith('Heading'):
            yield paragraph


def iter_block_items(parent):
    if isinstance(parent, _Document):
        parent_elm = parent.element.body
    elif isinstance(parent, _Cell):
        parent_elm = parent._tc
    elif isinstance(parent, _Row):
        parent_elm = parent._tr
    else:
        raise ValueError("something's not right")
    for child in parent_elm.iterchildren():
        if isinstance(child, CT_P):
            yield Paragraph(child, parent)
        elif isinstance(child, CT_Tbl):
            yield Table(child, parent)

**Comment by Aviru** [comment]
@nawuy  any help please.

**Comment by Aviru** [comment]
Hi,
Thank you for asking. Actually I am trying to build a resume parser. As,
resumes follow no specific format, so when I am parsing resumes, I am
seeing that in some resumes table head are top aligned and in some table
heads are left aligned reference the image I have posted in github. I am
trying to build a list which will contain key value pair for the tables
where the key will be table heads and value will be corresponding row or
column value  e.g [{'degree':'msc','name of institution':'abc'}] and so on.
But while traversing the table I am unable to differentiate between the
table heads and table body.
Thanks & Regards,
Aviru Bhattacharjee

**Comment by Ash921994** [comment]
> @BeebBenjamin if you install python-docx (v0.3.0+) rather than docx (v0.2.x) there is API I think you can use for this, roughly like so:
> 
> ```python
> document = Document(path_to_your_docx)
> tables = document.tables
> for table in tables:
>     for row in table.rows:
>         for cell in row.cells:
>             for paragraph in cell.paragraphs:
>                 print(paragraph.text)
> ```
> 
> To get the right version installed you might need to do this:
> 
> ```shell
> $ pip uninstall docx
> $ pip install -U --pre python-docx
> ```
> 
> The documentation you reference above is for this latest version. The legacy version doesn't have any documentation beyond the README.txt. Note the two versions are in separate repositories. This repository is for the latest one. The legacy version is here: https://github.com/mikemaccana/python-docx
> 
> Let me know if it gives you trouble.

**Comment by Ash921994** [comment]
Hi Scanny,

My docx file has paragraphs to be read in between the tables, but when I tried your example it doesn't work with my file. Can you please advise?
document = Document(path_to_your_docx)
tables = document.tables
for table in tables:
    for row in table.rows:
        for cell in row.cells:
            for paragraph in cell.paragraphs:
                print(paragraph.text)

Regards,
Ashish
